drivers/overlay/mount.go

```diff
@@ -141,14 +141,27 @@ func mountOverlayFromMain() {
 	// the new value for the list of lowers, because it's shorter.
 	if lowerv != "" {
 		lowers := strings.Split(lowerv, ":")
+		var newLowers []string
+		dataOnly := false
 		for i := range lowers {
+			if lowers[i] == "" {
```

Comment:
Non-blocking: If I’m following `composeFsLayersLowerDirs` construction correctly, this
- never happens twice in a row
- never happens for the last element.

If so,
```go
for i, oldLower := range lowers {
   if oldLower != "" {
      // … unix.Open …
      lowers[i] = fmt.Sprintf(…)
  }
}
```
or
```go
for _, lower := range lowers {
  if lower != "" {
      // … unix.Open …
      lower = fmt.Sprintf(…)
  }
  newLowers = append(newLowers, lower)
}
```

would be easier for me to follow.

The way I read the current `dataOnly` logic is “if this more complex logic than a trivial `map` is needed, it probably exists to handle consecutive, or trailing, empty entries”… but AFAICS that isn’t possible. Am I misreading the code?

Warning: I didn’t take the time to understand what the syntax / semantics of the value actually is…

Reply:
overlay now supports something like: `a:b::c::d` each directory that has `::` is considered a data-only layer (and at the moment regular layers cannot follow data-only paths but this could be relaxed in future), so `::` can happen multiple times.

Reply:
Oh, so it’s all my ignorance. Sorry about that.

Reply:
… and I was imprecise. “multiple times” is not a problem.

I was reading the logic in this PR to turn `a::::b:` into `fd1:fd2` (collapsing a _sequence_ of `:::` or more into a single `::`, and dropping the trailing one).

If that’s not relevant/expected, and it’s just a special `::`_something_ syntax, where preserving the original number/position of `:` chars is the right thing to do, then, yes, I think one of the two proposed loop alternatives (which don’t carry a value across iterations) would be a bit easier to follow.

Again, this is all non-blocking either way.